5 getting started to vitis acceleration flow with zynq 7000 (#82)
asynx-dev/www: docs/blog/20/5-gettig_started_to_vitis_acceleration_flow_with_zynq_7000.md

---
date: 2020-12-07
author: Anıl TIRLIOĞLU
description: Getting Started to Vitis Acceleration Flow with Zynq 7000.
axlang: "en"
---

# Getting Started to Vitis Acceleration Flow with Zynq 7000

> We all know, sometimes it is just hard to get start learning things from

Reviewer: ```suggestion
We all know, sometimes it is just hard to get started learning things from
```

vendor documentations. This tutorial will follow beginner friendly steps to run
your first accelerator on an FPGA.

> !!! Info

Reviewer: Admonition'lar bu şekilde çalışıyor gibi ama küçük harflerle tanımlanmışlar ilgili yerlerde. Gelecek sorun çıkartmamak adına küçük harfle, mesela `info` yapabilir miyiz? Yazının diğer kısımlarındakine ayrıca yorum yazmıyorum. Onları da düzeltebilir miyiz?

Ref:

- https://python-markdown.github.io/extensions/admonition/
- https://squidfunk.github.io/mkdocs-material/reference/admonitions/

    This tutorial will be more beneficial when used in
    conjunction with the official Xilinx tutorial[^1f] published on github.

There are two different device definitions in Xilinx. One is **embedded

> devices**[^2f],  other one is **data center accelerator cards**[^3f]. Embedded

Reviewer 2: Buna benzer satırlarda 80 kuralı uygulanırken eğer otomatik yapıyorsan "**embedded" gibi kısımlarda "**" ifadesi yarıda kesildiği için kalınlaştırılması istenilen yerler aslında kalınlaşmıyor bazen. Bu gibi kısımlar düzeltilmeli.

Reviewer: Bu benim de dikkatimi çekmişti. VSCode bu durumda editör içerisinde düzgün göstermiyor, haklısın. Fakat mkdocs ile siteyi derleyince düzgün gözüküyor. Bakabildiğim kadarıyla da Markdown speclerinde de izin verilen bir şey bu.

Github, VSCode gibi yerlerde düzgün gözükmesi için birleştirilebilir fakat sanıyorum bıraksak da çıkan sonuç düzgün olacak.

Reviewer 2: Bilgi için teşekkürler.

Reviewer: Ne demek :) markdownlint de hata vermeyince ona da güvendim biraz.

> device can work standalone. It hosts CPU and FPGA. On the other hand, data

Reviewer: ```suggestion
device can work standalone. It contains CPU and FPGA. On the other hand, data
```

> center accelerator cards needs to be attached to PC through PCIe interface. CPU

Reviewer: ```suggestion
center accelerator cards need to be attached to PC through PCIe interface. CPU
```

on your host PC will manage these FPGA cards. This tutorial will cover embedded
device flow with a Zynq 7000 series FPGA board ZC706.

!!! info
    Note that Xilinx tutorial targets Zynq Ultrascale+ ZCU102 and Alveo U200

>     boards for embedded flow and PCIe attached flow respectively. We will follow

Reviewer: ```suggestion
    boards for embedded flow and PCIe attached flow, respectively. We will follow
```

    embedded flow with ZC706[^4f].

> ZC706 board XC7Z045 FFG900 Zynq SoC chip. XC7Z045 is a midrange powerful 7

Reviewer: ```suggestion
ZC706 board has XC7Z045 FFG900 Zynq SoC on it. XC7Z045 is a midrange powerful 7
```

> series FPGA. Resource capacity is pretty enough to hands on tests with accelerators.

Reviewer: ```suggestion
series FPGA. Resource capacity is pretty enough to do hands on tests with accelerators.
```

| Resouce            | Quantity             |
| ------------------ | -----------          |

> | Logic Cells        | 350                  |

Reviewer: 350k olabilir mi? 350 yanlış scale'de gibi

| Block RAM (Mb)     | 19.1                 |
| DSP Slices         | 900                  |

Some essential terms could be helpful to know before starting to Vitis
acceleration flow.

| Term               | Description                                                                                                                  |
| ------------------ | -----------                                                                                                                  |
| Host               | CPU that runs main function, handles management and other possible sequential tasks, functions etc.                          |
| Device             | FPGA or board that accelerates functions which are called from host, mostly used in OpenCL definitions for GPU and FPGA      |

> | Kernel             | Region on FPGA that runs functions implemented for FPGA programmable logic, it is also called computing unit                 |

Reviewer: Kernel yerin adı mı yoksa işin adı mı? Alanı mı gösteriyor yoksa hızlandırıcı parçasını mı? Ben tam bilmiyorum, yorum yazarak kapatabilirsin bu görüşü, değiştir diye demiyorum :rabbit:

Author: birbirinden bağımsız düşünmüyorum. region içinde koşan fonksiyon sabit olduğu için.

Reviewer: Partial, DFX ile vs karışır mı diye düşünmüştüm de, oki. :+1:

| Accelerator        | Used interchangeably with kernel and device, in general it provides offloading of CPU                                        |

## Vector Add Example

The example we will run in this tutorial is a vector addition example. Two

> vectors that are build up with 4096 random numbers will be  vector summed on both

Reviewer: ```suggestion
vectors that are filled with 4096 random numbers will be  vector summed on both
```

> host which is Zynq Cortex A9 CPU and in vector_add kernel which is programmed to

Reviewer: ```suggestion
host which is Zynq Cortex A9 CPU and in `vector_add` kernel which is programmed in
```

> Zynq programmable logic. Results from device and from host will be then compared

Reviewer: ```suggestion
Zynq programmable logic. Results from the device and the from host will be then compared
```

> within host. Success and fail status will be returned on terminal screen of host.

Reviewer: ```suggestion
on the host. Success and fail status will be returned on terminal screen of the host.
```

## Required Tools and Files

### 0. GNU/Linux Operating System

This tutorial will be on Ubuntu operating system. Vitis documentation[^5f] lists
the supported other OSes as shown below.

![vitis_os_req](img/5-vitis_os_req.png)

You can view your Ubuntu version on console with `cat /etc/os-release`. I prefer
to use Ubuntu 16.04. Note that Ubuntu version 16.04.7 is not among supported OSes.
Vitis installation will be mentioned in the next step freezes at some point of
installation for this reason.

```console
$ cat /etc/os-release
NAME="Ubuntu"
VERSION="16.04.7 LTS (Xenial Xerus)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 16.04.7 LTS"
VERSION_ID="16.04"
HOME_URL="http://www.ubuntu.com/"
SUPPORT_URL="http://help.ubuntu.com/"
BUG_REPORT_URL="http://bugs.launchpad.net/ubuntu/"
VERSION_CODENAME=xenial
UBUNTU_CODENAME=xenial
```

To trick the installer, set your OS version to one of the closest supported one
with your OS. Open with the text editor with command below and change `16.04.7`
to  `16.04.6`.

```console
$ sudo gedit /etc/os-release
```

Now, make sure also that you have required software packages or run the command below
to install them.

```console

> $ sudo add-apt-repository ppa:xorg-edgers/ppa sudo apt-get updatesudo apt-get install libgl1-mesa-glxsudo apt-get install libgl1-mesa-drisudo apt-get install libgl1-mesa-devsudo add-apt-repository --remove ppa:xorg-edgers/ppasudo apt install net-toolssudo apt-get install -y unzipsudo apt install gccsudo apt install g++sudo apt install pythonln -s /usr/bin/python2 /usr/bin/python

Reviewer: Bu kod çalışıyor mu yapıştırınca? Tek satrda arka arkaya bir sürü `sudo` var :stop_sign:  Ayrı ayrı mı `$` olmalı ya da bir adet `apt install`.
`install g++sudo apt install pythonln` zaten `g++` ile `sudo` birleşmiş. Buralar sıkıntılı komple sanki

`pythonln -s /usr/bin/python2 /usr/bin/python` bu ayrı komut mu?

```

You can now run the installer in the following step.

### 1. Vitis

> Go Xilinx download page <https://www.xilinx.com/support/download.html>

Reviewer: ```suggestion
Go to the Xilinx download page <https://www.xilinx.com/support/download.html>
```

I prefer downloading **Linux Self Extracting Web Installer**.

![Vitis_Download](img/5-vitis_download.png)

> Make the bin file executable ```$ chmod +x Xilinx_Unified_2020.1_0602_1208_Lin64.bin```

Reviewer: İnline kısımlarda 3 backquote sanıyorum anlamlı olmuyor. Code block olarak diğer yerlerdekiler anlamlı da inline kısımlarda tek olması daha uygun sanıyorum.

and run the installer `$ ./Xilinx_Unified_2020.1_0602_1208_Lin64.bin`.

Follow the installation defaults. After the installation; Vivado, Vitis and other

> xilinx tools should be under `/opt/tools/Xilinx/` directory.

Reviewer: ```suggestion
Xilinx tools should be under `/opt/tools/Xilinx/` directory.
```

> ### 2. Xilinx Runtime(XRT)

Reviewer: ```suggestion
### 2. Xilinx Runtime (XRT)
```

> Go Xilinx download page <https://www.xilinx.com/support/download.html> and open

Reviewer: ```suggestion
Go to the Xilinx download page <https://www.xilinx.com/support/download.html> and open
```

see `/opt/xilinx/xrt/` directory created and **XRT** is installed.

### 3. Embedded Platform Files for ZC706

Go Xilinx download page <https://www.xilinx.com/support/download.html> and open
`Vitis Embedded Platforms` tab again.

![Platform_Download](img/5-platform_download.png)

Download ZC706 base platform zip file. Extract the zip file to location `/opt/xilinx/platforms/`.
Now, you should have file `xilinx_zc706_base_202010_1` under this directory.

### 4. Prebuilt Petalinux Kernel and Root File System

We will use prebuilt kernel and rootfs for simplicity. Go Xilinx download page
<https://www.xilinx.com/support/download.html> and open `Petalinux` tab.

![Common_Download](img/5-zynq_common_download.png)

Download `ZYNQ Common Image` and extract it to `/opt/tools/Xilinx/xilinx-zynq-common-v2020.1/`.
Run `sh sdk.sh` script again in this folder. This script should create **sysroot**
under `/opt/petalinux/2020.1/`.

### 5. Tutorial Files

In your home directory create `projects` folder to download tutorial files there.

* Within `projects`, clone tutorial files from github repository `$ git clone https://github.com/aniltirli/Vitis-Tutorials.git`

> 

Reviewer: Burada neden senden clone'luyoruz :smiley: `patch-1` in bir anlamı var mı?

Author: normalde zc706 yok, zc706 için eklemeler yaptım.

* Tutorial files should reside under `Vitis-Tutorials/Getting_Started/Vitis/example/zc706`.

## Setting Environmental Variables

We need to configure Vitis environmental settings to be able to use required
tools and scripts. Run the commands below.

```console
$ source /opt/tools/Xilinx/Vitis/2020.1/settings64.sh
$ source /opt/xilinx/xrt/setup.sh
$ unset LD_LIBRARY_PATH
$ source $XILINX_VITIS/data/emulation/qemu/unified_qemu_v4_2/environment-setup-aarch64-xilinx-linux

$ export LIBRARY_PATH=/usr/lib/x86_64-linux-gnu

$ export PLATFORM_REPO_PATHS=/opt/xilinx/platforms
$ export ROOTFS=/opt/tools/Xilinx/xilinx-zynq-common-v2020.1
$ export SYSROOT=/opt/petalinux/2020.1/sysroots/cortexa9t2hf-neon-xilinx-linux-gnueabi
$ source /opt/petalinux/2020.1/environment-setup-cortexa9t2hf-neon-xilinx-linux-gnueabi
```

!!! tip
    Commands above lasts only for a console session. Create script file
    `vitis_settings.sh` and copy commands above there. Whenever you need to set
    environment for Vitis acceleration flow you can run this script `$ sh vitis_settings.sh`.

## Running Emulations

These steps are similar to steps mentioned in Xilinx tutorial embedded flow.

### 1. Software Emulation

> Software emulation will run on QEMU virtualizing Zynq on your Ubuntu. This simulation

Reviewer: ```suggestion
Software emulation will run on QEMU virtualizing Zynq on your Ubuntu. This emulation
```

will provide fast error detection for syntax and behavioral errors. Hardware

> accelerator code(`Vitis-Tutorials/Getting_Started/Vitis/example/src/vadd.cpp`)

Reviewer: ```suggestion
accelerator code (`Vitis-Tutorials/Getting_Started/Vitis/example/src/vadd.cpp`)
```

> and host code(`Vitis-Tutorials/Getting_Started/Vitis/example/src/host.cpp`) will

Reviewer: ```suggestion
and host code (`Vitis-Tutorials/Getting_Started/Vitis/example/src/host.cpp`) will
```

be compiled for only Zynq Cortex ARM A9 processor.

Under the directory
`Vitis-Tutorials/Getting_Started/Vitis/example/zc706/sw_emu/`,
you should find `build_and_run.sh`, run this script ```$ sh build_and_run.sh```.

After QEMU boots run the commands below.

```console
$ mount /dev/mmcblk0p1 /mnt
$ cd /mnt
$ cp platform_desc.txt /etc/xocl.txt
$ export XILINX_XRT=/usr
$ export XILINX_VITIS=/mnt
$ export XCL_EMULATION_MODE=sw_emu
$ ./app.exe
```

![sw_emu](img/5-sw_emu.png)

!!! Warning
    If you get `ERROR: parted command not exist` error, add location of parted
    to PATH export ```$ export PATH=$PATH:/sbin``` or install parted

>     ```$ sudo apt install parted```.

Reviewer: Yukarıda yazdığım 3 lü backquote sanıyorum github markdown render'ingi kafasını karıştırıyor. Bu sayfada aşağısının renkleri kaymış bir miktar.

### 2. Hardware Emulation

This simulation will take long but will provide cycle accurate performance and
profiling without needing to real physical hardware. Hardware accelerator code
and host code is compiled for Zynq programmable logic side and Zynq ARM Cortex

> A9 processor respectively. Emulation will run on QEMU which is virtualizing Zynq

Reviewer: ```suggestion
A9 processor, respectively. Emulation will run on QEMU which is emulating a Zynq
```

QEMU booted, execute commands below on terminal screen.

```console
$ mount /dev/mmcblk0p1 /mnt
$ cd /mnt
$ cp platform_desc.txt /etc/xocl.txt
$ export XILINX_XRT=/usr
$ export XILINX_VITIS=/mnt
$ export XCL_EMULATION_MODE=hw_emu
$ ./app.exe
```

![hw_emu](img/5-hw_emu.png)

## Running Accelerator on ZC706

Under the directory
`Vitis-Tutorials/Getting_Started/Vitis/example/zc706/hw_emu/`,
you will find `build.sh` run this script `$ sh build.sh`. This may take time
to finish. `package` folder should be created by the script. ZC706 will boot
from SD card in this tutorial. Switch SW11 on ZC706 board should look like below
image for boot mode SD card.

![zc706_sw11](img/5-zc706_sw11.png)

### Creating SD card Image

Insert SD card to your PC and check its name with command.

```console
$ sudo fdisk -l
```

Most likely, SD card will be named **mmcblk0** and it's partitions will be printed
on console.

```console
$ sudo umount /dev/mmcblk0
```

After making sure it is the SD card you want to use, format and copy SD card image
***sd_card.img*** created with `build.sh` script. Write it to your SD card.

```console

> $ dd if=/dev/mmcblk0 of=./package/sd_card.img

Reviewer: `if` ve `of` ters gibi

            std::cout << "i = " << i << " CPU result = " << expected << " Device result = " << out[i] << std::endl;
            match = false;
            miss_counter++;
            //break;
        }
        std::cout << "i = " << i << " CPU result = " << expected << " Device result = " << out[i] << std::endl;
    }
    std::cout << miss_counter << " results mismatched" << std::endl;
```

You will not need to run scripts every time since you only changed `host.cpp`.
Compile the `app.exe` again with the command below inside `hw` folder.

```console
$ arm-linux-gnueabihf-g++ -Wall -g -std=c++11 ../../src/host.cpp -o app.exe -I${SYSROOT}/usr/include/xrt -L${SYSROOT}/usr/lib -lOpenCL -lpthread -lrt -lstdc++ --sysroot=${SYSROOT}
```

Copy `app.exe` to SD card attached to your board, with ssh from your PC.

```console
$ scp app.exe  root@<board_ip_address>:/mnt
```

You can run `app.exe` from board's console again.

```console
$ ./app.exe
```

!!! Warning
    Your board should have an IP address and should be connected to your PC
    over ethernet for ssh connection.

![after_change](img/5-after_change.png)

<!---Refs --->

> [^1f]:https://github.com/Xilinx/Vitis-Tutorials/tree/master/Getting_Started/Vitis

Reviewer: Önceki yazılarda da burada eksik bırakmışız. Böyle olunca linkler tıklanmıyor, eskileri için de issue açacağım. Bunları da yazı içindeki linkler gibi `<>` yapabilir misin? `:` dan sonra da bir boşluk atabilirsin.
